Consume Move pointer events in `TouchEvent`

L-Andrade opened the pull request:
When a graph without horizontal scroll is in a horizontally scrollable component, such as a HorizontalPager, the events are not being consumed, resulting in a scroll when the user is interacting with the graph.
The `TouchEvent` should consume the move events so that the events are not processed by other scroll/touch handlers.

This change is only applied when scrolling is not enabled. When scrolling is enabled, the behaviour is the same.

Here's a before (old) and after (new) video with the change.

https://github.com/user-attachments/assets/030d09d6-e43f-4c7b-8e77-14dccee4d53b

https://github.com/user-attachments/assets/700e2d49-a603-4de5-80de-08a9c48c08e0

Please let me know if I missed anything, as this might be on purpose.

Gowsky commented:
Hello, @L-Andrade. This has been the subject of several threads. Previously, gestures were consumed in this case. This behavior was [reported as a bug](https://github.com/patrykandpatrick/vico/issues/942) and replaced with the one that's currently in place. After this replacement, the new behavior was also [reported as a bug](https://github.com/patrykandpatrick/vico/issues/985).

Since either behavior has its fans, it seems that instead of changing the consumption logic once again, we should make this customizable, as mentioned [here](https://github.com/patrykandpatrick/vico/issues/985#issuecomment-2568280069). Are you interested in adding such a parameter? (It would have to be added for all three frameworks: Jetpack Compose, Compose Multiplatform, and views.)

In the meantime, [this workaround](https://github.com/patrykandpatrick/vico/issues/985#issuecomment-2570332828) may be helpful.

L-Andrade commented:
Hey @Gowsky 

Thanks for the reply! I'm definitely interested in doing so. 

I'll take a look when I have the time, not sure when that will be.

I'll tag you here if I have questions regarding the implementation.

L-Andrade commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/TouchEvent.kt:
```diff
@@ -27,15 +27,14 @@ import com.patrykandpatrick.vico.core.common.Point
 
 internal fun Modifier.chartTouchEvent(
   setTouchPoint: ((Point?) -> Unit)?,
-  isScrollEnabled: Boolean,
```
I've removed this from the arguments since it's already in `scrollState` which is also passed in another argument

L-Andrade commented on vico/views/src/main/java/com/patrykandpatrick/vico/views/cartesian/CartesianChartView.kt:
```diff
@@ -106,12 +107,15 @@ constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
   public var scrollHandler: ScrollHandler by
-    invalidatingObservable(ScrollHandler(themeHandler.scrollEnabled)) { oldValue, newValue ->
+    invalidatingObservable(
+      ScrollHandler(themeHandler.scrollEnabled, themeHandler.consumeMoveEvents)
```
I've followed the implementation of `scrollEnabled` for the new `consumeMoveEvents`, but I'm not sure if this is the best choice

L-Andrade reviewed (commented):
Hello @Gowsky 
In the last commit (7bdbe27f912ae54e7849b08945a7880864f0e2ae), I've added a custom property to `VicoScrollState`: `consumeMoveEvents`.

Let me know if I missed anything, or if the new logic does not meet the existing implementation!

I have a video of the behaviour, but please take a look yourself too, of course:

https://github.com/user-attachments/assets/52549e23-1c4c-486a-8741-b8052eeb4040

Gowsky reviewed (changes requested):
Thanks for the changes. We have some suggestions:
- `consumeMoveEvents` shouldn't be in `VicoScrollState` or `ScrollHandler`, since it doesn't relate to scrolling.  Move gestures are used not just for scrolling, but also for dragging the marker. Thus it should be a parameter of `CartesianChartHost`, and a property of `CartesianChartView`.
- `consumeMoveEvents` is not needed in `MeasuringContext`.
- These changes need to be applied in the `multiplatform` module as well.

L-Andrade commented:
> Thanks for the changes. We have some suggestions:
> 
> - `consumeMoveEvents` shouldn't be in `VicoScrollState` or `ScrollHandler`, since it doesn't relate to scrolling.  Move gestures are used not just for scrolling, but also for dragging the marker. Thus it should be a parameter of `CartesianChartHost`, and a property of `CartesianChartView`.

Oh, that's fair. I've adjusted this in: a939fd3656b671a24202bb891dea334816c67439

> -  `consumeMoveEvents` is not needed in `MeasuringContext`.

Yeah, my bad! The view stuff threw me off a bit. I also addressed this in a939fd3656b671a24202bb891dea334816c67439

Let me know if this looks better. I've tested it briefly in Compose and Compose multiplatform, and both seemed to work as expected. Here's a short demo:

https://github.com/user-attachments/assets/6f66ba9f-b441-48bd-aa2d-00bd0ef79a60

Thanks for the feedback, and sorry for the delay!

Gowsky commented on vico/views/src/main/java/com/patrykandpatrick/vico/views/cartesian/CartesianChartView.kt:
```diff
@@ -325,6 +325,7 @@ constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
       if (chart.layerBounds.isEmpty) return@withChartAndModel
 
       motionEventHandler.scrollEnabled = scrollHandler.scrollEnabled
+      motionEventHandler.consumeMoveEvents = themeHandler.consumeMoveEvents
```
It should also be possible to toggle consumption programmatically. Thus, instead of updating  `consumeMoveEvents` with `themeHandler.consumeMoveEvents` here, a `consumeMoveEvents` property should be added to `CartesianChartView`. It should delegate the setter and the getter to `motionEventHandler.consumeMoveEvents`. `MotionEventHandler` should have a new constructor parameter for `consumeMoveEvents` which would be set using the `themeHandler`.

L-Andrade replied:
Makes sense. I've adjusted the constructor and added the changes in this commit: a92786e4ecf216f43b49cd44184010a6f27426f1

Let me know if anything's missing!

Gowsky replied:
Everything's good now, thanks! We'll make some minor tweaks soon and merge the PR.

Gowsky reviewed (changes requested):
There's just one thing left that needs to be changed. After that, we will make minor tweaks and merge it.

v01dn3ph1l1m commented:
@Gowsky when can we expect the merge? has been a while since I have been following this thread. 
Awaiting a release with this feature :)

Thanks a bunch for this library. its the best out there for compose multiplatform

Gowsky commented:
@L-Andrade, thank you for your contribution.

Hello, @v01dn3ph1l1m. We appreciate your kind words. The PR has just been merged, thus I think this answers your question. Cheers!

v01dn3ph1l1m commented:
@Gowsky thank you. can we have a version release for this?

shervinkoushan commented:
> @Gowsky thank you. can we have a version release for this?

Second that, would be great with a release @Gowsky :)

Gowsky commented:
[Vico `2.2.0-alpha.1`](https://github.com/patrykandpatrick/vico/releases/tag/v2.2.0-alpha.1) includes this change.
